### ISSUE TITLE
ofAVFoundationPlayer: isLoaded returns false after a movie loaded for a short time

### ISSUE THREAD

**Issue by r9y9** [openframeworks/openFrameworks#4629]
Even if a movie is successfully loaded synchronously, the `isLoaded` function returns false for a short time, and then return true.  I wonder it should always return true after a movie is loaded, but is this normal behavior or a bug?

## Code

In ofApp.h: 
```cpp
 ofVideoPlayer videoPlayer;
```

In ofApp.cpp:
```cpp
//--------------------------------------------------------------
void ofApp::setup() {
    // sample video from:
    // http://www.sample-videos.com/video/mp4/720/big_buck_bunny_720p_1mb.mp4
    bool success = videoPlayer.load("SampleVideo_1080x720_1mb.mp4");
    ofLogNotice() << ofGetTimestampString() << ", load: " << success;
    ofLogNotice() << ofGetTimestampString()
                  << ", isLoaded: " << videoPlayer.isLoaded();

    videoPlayer.play();
}

//--------------------------------------------------------------
void ofApp::update() {
    ofLogNotice() << ofGetTimestampString()
                  << ", isLoaded: " << videoPlayer.isLoaded();
    videoPlayer.update();
}

//--------------------------------------------------------------
void ofApp::draw() { videoPlayer.draw(0, 0); }
```

## Results

```
2015-11-27 17:49:30.619 videoPlayerExampleDebug[53520:3446574] video loaded at 1280 x 720 @ 25.000000 fps
[notice ] 2015-11-27-17-49-30-626, load: 1
[notice ] 2015-11-27-17-49-30-626, isLoaded: 0
[notice ] 2015-11-27-17-49-30-626, isLoaded: 0
[notice ] 2015-11-27-17-49-30-811, isLoaded: 0
[notice ] 2015-11-27-17-49-30-816, isLoaded: 0
[notice ] 2015-11-27-17-49-30-819, isLoaded: 0
[notice ] 2015-11-27-17-49-30-917, isLoaded: 0
[notice ] 2015-11-27-17-49-30-934, isLoaded: 0
[notice ] 2015-11-27-17-49-30-935, isLoaded: 0
[notice ] 2015-11-27-17-49-31-022, isLoaded: 1
[notice ] 2015-11-27-17-49-31-031, isLoaded: 1
[notice ] 2015-11-27-17-49-31-067, isLoaded: 1
[notice ] 2015-11-27-17-49-31-068, isLoaded: 1
[notice ] 2015-11-27-17-49-31-164, isLoaded: 1
[notice ] 2015-11-27-17-49-31-194, isLoaded: 1
[notice ] 2015-11-27-17-49-31-207, isLoaded: 1
[notice ] 2015-11-27-17-49-31-220, isLoaded: 1
[notice ] 2015-11-27-17-49-31-259, isLoaded: 1
[notice ] 2015-11-27-17-49-31-277, isLoaded: 1
[notice ] 2015-11-27-17-49-31-281, isLoaded: 1
[notice ] 2015-11-27-17-49-31-286, isLoaded: 1
...
```

As you can see, there are a few zeros after a movie loaded.

## Environment

- Mac OS X 10.10.4
- oF current master f2149a97c85154a531c1395f21e29f257d67901b
- debug app for 64bit

**Comment by r9y9** [comment]
After looking the source of AVFoundationVideoPlayer, I'm convinced that this behavior is possible to happen... This might be a API change (since that works synchronously as expected before on of 0.7.3 at least. sorry for not upgrading for a while) rather than a bug, but it seems like the current behavior of the `load` function is asynchronous. Not sure the `load` function should behave completely synchronously or not.

probably related: #3750, http://forum.openframeworks.cc/t/dealing-with-asynchronous-calls/19529

**Comment by ofTheo** [comment]
Thanks for reporting this issue. 

The reason this is happening is the AVF based video player is returning the isReady bool not the isLoaded bool to the ofVideoPlayer. 

I believe this was done so you don't try and access pixels before the internal AVF player actually has pixels for you which can sometimes be a few milliseconds after the video is loaded. 

For me using isLoaded instead - corrects the issue on the surface, but there are some internal warnings that get thrown relating to unallocated textures because the player believes there is something to show. 

This wouldn't be an issue if the videos were truly loading synchronously. 
Synchronous loading would imply that as soon as .load() returns the video should be usable, which it is not currently. 

An easy fix would be to do something like this:

pseudo code:
```
if( aSync == false ){ 
  while( movie.isLoaded && movie.isReady == false ){
      //wait for loading.    
  }
}
```

With some sort of timeout to handle a case where movie.isReady doesn't return true. 
I'll give this a go quickly and see if it solves the issue and makes sync loading actually ready after load.

**Comment by ofTheo** [comment]
hmm for some reason trying to block until the movie isReady is stopping AVF internally from making the movie ready, I'll do some more research on this today, it would be great to get sync loading working properly.

**Comment by r9y9** [comment]
Yeah, I tried the same easy fix before and was unable to solve the problem. Since I have almost no experience on AVFoundation, I'm really appreciate it if you or someone who is familiar with AVFoundation could investigate the problem. Thanks!

**Comment by ofTheo** [comment]
So it looks like there are two things here which are going on. 
The synchronous loading is happening correctly - the code waits till the file is loaded before returning. 

However there is a short period between when the video is loaded and the callback from the player reports AVPlayerItemStatusReadyToPlay. 

This is where the bReady bool is set true and this is actually when you can start using the pixels. 
I'm trying to see if there is a way to block until AVPlayerItemStatusReadyToPlay is true. 

Everything I have tried to wait for the AVPlayerItemStatusReadyToPlay status stops the observeValueForKeyPath callback from getting called.

Going to ping @i-n-g-o @julapy and @openframeworks/video

**Comment by ofTheo** [comment]
@r9y9 the above PR should be a fix to the issue you were having. 

There isn't a good way to block until pixels are ready - which I don't think really is probably a good idea as it could cause more issues than it solves.  

But if you load a movie you can query isLoaded() to see if it is loaded successfully. ( in your example it is now returning true as soon as load() returns ) 

Thanks for reporting this issue!
Theo

**Comment by r9y9** [comment]
@ofTheo Thanks for tackling the issue! I tried your bug fix branch (#4737), however, I got another issue; `isFrameNew()` always returns false, even if `videoPlayer.update()` calls. Perhaps internal texture isn't updated correctly?

Modified code (setup, update are same as in this issue description):

```cpp
//--------------------------------------------------------------
void ofApp::draw() {
    if (videoPlayer.isFrameNew()) {
        videoPlayer.draw(0, 0);
    } else {
        ofLogNotice() << "Frame is not ready";
    }
}
```

result:

```
2016-01-18 12:23:04.439 videoPlayerExampleDebug[71654:2242649] video loaded at 1280 x 720 @ 25.000000 fps
[notice ] 2016-01-18-12-23-04-448, load: 1
[notice ] 2016-01-18-12-23-04-448, isLoaded: 1
[notice ] 2016-01-18-12-23-04-448, isLoaded: 1
[notice ] Frame is not ready
[notice ] 2016-01-18-12-23-04-674, isLoaded: 1
[notice ] Frame is not ready
[notice ] 2016-01-18-12-23-04-684, isLoaded: 1
[notice ] Frame is not ready
[notice ] 2016-01-18-12-23-04-696, isLoaded: 1
[notice ] Frame is not ready
[notice ] 2016-01-18-12-23-04-798, isLoaded: 1
[notice ] Frame is not ready
...
```

**Comment by r9y9** [comment]
Ok, a reduced test case:

```cpp
//--------------------------------------------------------------
void ofApp::update() { videoPlayer.update(); }

//--------------------------------------------------------------
void ofApp::draw() { videoPlayer.draw(0, 0); }
```

The code above always shows the same texture (probably the first frame of the movie).
